velox/dwio/common/tests/TestParallelBufferedInput.cpp

```diff
@@ -0,0 +1,190 @@
+    stream->readFully(buffer->data(), region.length);
+
+    // Verify data correctness.
+    auto expect = readFile->pread(region.offset, region.length);
+    ASSERT_EQ(buffer->size(), expect.size());
+    ASSERT_EQ(buffer->data(), expect);
+  }
+
+  int32_t totalLength = 0;
+  for (const auto& region : regions) {
+    totalLength += region.length;
+  }
+  ASSERT_EQ(totalLength, ioStats->rawBytesRead());
+
+  ASSERT_EQ(0, ioStats->rawOverreadBytes());
+}
+
+class TestReadFile : public ReadFile {
+ public:
+  explicit TestReadFile(std::string_view file) : file_(file) {}
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    // Assume read latency is 1 millisecond per MB.
+    int64_t latency = length >> 20;
+    std::this_thread::sleep_for(std::chrono::milliseconds(latency));
+    return {static_cast<char*>(buf), length};
+  }
+  bool shouldCoalesce() const override {
+    return false;
+  }
+  uint64_t size() const override {
+    return 1 << 30;
+  }
+  uint64_t memoryUsage() const override {
+    return 0;
+  }
+  std::string getName() const override {
+    return "dummy";
+  }
+  uint64_t getNaturalReadSize() const override {
+    return 1 << 20;
+  }
+
+ private:
+  std::string_view file_;
+};
+
+TEST(TestParallelBufferedInput, simpleBenchmark) {
+  std::string filename = "test-file";
+  const int32_t kOneMB = 1 * 1024 * 1024;
+  const int32_t threads = 4;
+  int32_t loadQuantum = 8 * kOneMB;
+  auto executor = std::make_unique<folly::IOThreadPoolExecutor>(threads);
+  auto pool = memory::getDefaultMemoryPool();
+  auto readFile = std::make_shared<TestReadFile>(filename);
+
+  std::vector<Region> regions;
+
+  // 10 regions, each has 2MB length and 2MB distance, eg [0, 2MB], [4MB,
+  // 2MB]...
+  for (uint64_t iter = 0, cursor = 0, length = 2 * kOneMB; iter < 10;
+       ++iter, cursor += 4 * kOneMB) {
+    regions.emplace_back(cursor, length);
+  }
+  // 20 regions, each has 14MB length and 2MB distance, eg [0, 14MB], [16MB,
+  // 14MB]...
+  for (uint64_t iter = 0, cursor = 0, length = 14 * kOneMB; iter < 20;
+       ++iter, cursor += 16 * kOneMB) {
+    regions.emplace_back(cursor, length);
+  }
+
+  auto stats = std::make_shared<IoStatistics>();
+  auto parallelStats = std::make_shared<IoStatistics>();
+  auto parallelFunc = [&]() {
+    ParallelBufferedInput parallelInput(
+        readFile,
+        *pool,
+        MetricsLog::voidLog(),
+        parallelStats,
+        executor.get(),
+        loadQuantum);
+    for (const auto& region : regions) {
+      parallelInput.enqueue({region.offset, region.length});
+    }
+    parallelInput.load(LogType::TEST);
+  };
+  auto normalFunc = [&]() {
+    BufferedInput input(readFile, *pool, MetricsLog::voidLog(), stats.get());
+    for (const auto& region : regions) {
+      input.enqueue({region.offset, region.length});
+    }
+    input.load(LogType::TEST);
+  };
+
+  // Execute 4 times test in alternating order, compare ioStats.
```

Comment:
What is the motivation behind alternating the order and repeating them 4 times in this test?
I think one should be sufficient since different IoStatistics objects are used.

Reply:
To avoid system cache or other potential cache outside Velox, not related with IoStatistics objects.